src/test/py/bazel/genrule_test.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import unittest
+from src.test.py.bazel import test_base
+
+
+class GenRuleTest(test_base.TestBase):
```

Comment:
You should also assert quoting -- what if an input or output file (`$@`) has a space? (`$(location)` also supports spaces.)

See also https://github.com/bazelbuild/bazel/issues/6531

Reply:
I checked `$@` won't add any quotes if the path contains spaces, but `$(location)` does add single quote in that case. Single quote will work in powershell but not in cmd.exe. However, I don't think that's a behavior related to my change and we already have tests for that in other places.  See https://github.com/bazelbuild/bazel/commit/aacc305ba5a2035155cdc9afc9a300c739282829

Reply:
Thanks